File: CSharpPools/PoolsLibrary/Controller/PoolController.cs
using System;
using System.Collections.Concurrent;
using System.Threading;
using PoolsLibrary.Pool;

namespace PoolsLibrary.Controller
{
    public class PoolController<TK, TV> : IDisposable
    {
        internal ConcurrentDictionary<TV, TK> ObtainedObjectToItsKey
        {
            get { return _obtainedObjectToItsKey; }
        }

        private readonly PoolControllerSettings _settings;
        private readonly ConcurrentDictionary<TV, TK> _obtainedObjectToItsKey = new ConcurrentDictionary<TV, TK>();
        private bool _disposingWasCalled = false;

        private readonly IPool<TK, TV> _pool;

        public PoolController(PoolControllerSettings settings, IPool<TK, TV> pool)
        {
            _settings = settings;
            _pool = pool;
        }

        public bool Obtain(TK key, out TV outPoolObject, DirectionIfNoObjectIsAvailable<TK, TV> noObjectDirection)
        {
            if (_disposingWasCalled)
            {
                outPoolObject = default(TV);
                return false;
            }

            var adaptedDirection = AdaptDirection(noObjectDirection);

            int curAttemptNumber = 0;
            while (++curAttemptNumber <= adaptedDirection.AttemptsNumber)
            {
                if (curAttemptNumber > 1)
                    Thread.Sleep(adaptedDirection.OneIntervalBetweenAttemptsInSeconds * 1000);

                try
                {
                    var curAttemptDelegate = ChooseCreateDelegate(adaptedDirection, curAttemptNumber);
                    SharedEnvironment.Log.DebugFormat("Attempt #{0} with delegate='{1}'", curAttemptNumber, curAttemptDelegate);

                    if (_pool.TryObtain(key, out outPoolObject, curAttemptDelegate))
                    {
                        if (_settings.CallingReleaseOperationWillHappen)
                            ObtainedObjectToItsKey.TryAdd(outPoolObject, key);
                        return true;
                    }
                }
                catch (Exception e)
                {
                    var internalErrorMessage = string.Format("Something failed during attempt #{0} of obtaining object " +
                                                             "with key='{1}'. Look at inner exception for details",
                                                             curAttemptNumber, key);
                    throw new PoolException<TK>(key, internalErrorMessage, e);
                }
            }

            outPoolObject = default(TV);
            return false;
        }

        public void Release(TV objectToRelease)
        {
            if (_disposingWasCalled)
                return;
            CheckPromiseToCallReleaseOperation(objectToRelease);

            TK key;
            if (ObtainedObjectToItsKey.TryRemove(objectToRelease, out key))
            {
                _pool.Release(key, objectToRelease);
                return;
            }

            var notObtainedObjectMessage = "Only obtained objects are allowed to be released";
            throw new InvalidPoolOperationException<TK, TV>(objectToRelease, message:notObtainedObjectMessage);
        }

        /// For performance reasons there is no synchronization between disposing and
        /// obtaining+releasing operatons in the internal pool components. So user is
        /// supposed to dispose controller only after completion of all other operations.
        /// Otherwise some resources may be not disposed or even pool consistency may be broken.
        /// PoolController helps an user a little bit: if some operation is called after
        /// starting disposing, it will be rejected (silently ignored).
        /// <summary>
        /// May be called only after completion of all obtaining and releasing operations.
        /// </summary>
        public void Dispose()
        {
            SharedEnvironment.Log.Debug("Disposing has started");
            _disposingWasCalled = true;

            _pool.Dispose();
            SharedEnvironment.Log.Debug("Disposing has finished");
        }

        private DirectionIfNoObjectIsAvailable<TK, TV> AdaptDirection
                                                (DirectionIfNoObjectIsAvailable<TK, TV> originalDirection)
        {
            var toReturn = originalDirection ?? DirectionIfNoObjectIsAvailable<TK, TV>.DoNotWaitDirection;

            if (!_settings.CallingReleaseOperationWillHappen)
                toReturn.AttemptsNumber = 1;

            return toReturn;
        }

        private Func<TK, TV> ChooseCreateDelegate(DirectionIfNoObjectIsAvailable<TK, TV> direction,
                                          int curAttemptNumber)
        {
            var isLastAttempt = curAttemptNumber == direction.AttemptsNumber;
            return isLastAttempt
                   ? direction.CreateDelegateIfNoObjectIsAvailable
                   : null;
        }

        private void CheckPromiseToCallReleaseOperation(TV objectToRelease)
        {
            if (_settings.CallingReleaseOperationWillHappen)
                return;

            var noPromiseMessage = "In order to release objects promise it by setting " +
                                   "PoolControllerSettings.CallingReleaseOperationWillHappen " +
                                   "to true. Currently it's false";
            throw new InvalidPoolOperationException<TK, TV>(objectToRelease, message:noPromiseMessage );
        }
    }
}

File: JavaPools/PoolsLibrary/src/main/java/ru/urfu/javapools/poolslibrary/controller/PoolController.java
package ru.urfu.javapools.poolslibrary.controller;

import java.util.WeakHashMap;

import ru.urfu.javapools.poolslibrary.function.FunctionThatMayThrow;
import ru.urfu.javapools.poolslibrary.locks.NotReentrantReadWriteLock;
import ru.urfu.javapools.poolslibrary.pool.IPool;
import ru.urfu.javapools.poolslibrary.pool.basicfunctionality.item.NoAvailableObjectException;
import ru.urfu.javapools.poolslibrary.pool.basicfunctionality.item.ObjectsMaxCountReachedException;

public class PoolController<TK,TV> implements AutoCloseable {
	
	private final IPool<TK,TV> _pool;
	
	private final WeakHashMap<TV,TK> _obtainedObjectToItsKey = new WeakHashMap<TV,TK>();	
	private NotReentrantReadWriteLock _obtainedObjectToKeyLock = new NotReentrantReadWriteLock();		

	private final static String _MESSAGE_PATTERN_OBTAINING_FAILED = "Something failed during attempt #%s of obtaining" +
																	" object with key='%s'. Look at cause for details";
	private final static String _MESSAGE_PATTERN_OBTAINING_INTERRUPTED = "Controller obtained object with key='%s'," +
																		 " but user thread was interrupted during finishing obtaining" +
																		 " - controller has released object";
	private final static String _MESSAGE_RELEASING_OF_STRANGER = "Only obtained objects are allowed to be released";
	
	public PoolController(IPool<TK,TV> pool) {
		_pool = pool;
	}
	
	public TV obtain (TK key, DirectionIfNoObjectIsAvailable<TK, TV> noObjectDirection)
		throws PoolException {
		
		TV toReturn = null;
		
		if (noObjectDirection == null)
			noObjectDirection = new DirectionIfNoObjectIsAvailable<TK,TV>().setAttemptsNumber(1); 
		
		int curAttemptNumber = 0;
		while (curAttemptNumber++ < noObjectDirection.getAttemptsNumber()) {	
		
			if (curAttemptNumber > 1)
				waitSafely(noObjectDirection.getOneIntervalBetweenAttemptsInSeconds(), key);			
			
			boolean isLastAttempt = curAttemptNumber == noObjectDirection.getAttemptsNumber();
			FunctionThatMayThrow<TK,TV> curDelegate = isLastAttempt ? noObjectDirection.getCreateMethod() : null;
			
			try {				
				toReturn = _pool.obtain(key, curDelegate);
				break;
			} catch (NoAvailableObjectException | ObjectsMaxCountReachedException e) {
				if (isLastAttempt)
					throw e;
			} catch (PoolException e) {
				String message = String.format(_MESSAGE_PATTERN_OBTAINING_FAILED, curAttemptNumber, key);
				throw new PoolException(key, message, e);
			}
		}

		rememberObtainedObject(key, toReturn);
		return toReturn;
	}
	
	public void release(TV objectToRelease) throws PoolException {
		
		TK key = null;
		
		try {
			key = getKeyByObject(objectToRelease);
		} catch (InterruptedException e) {
			Thread.currentThread().interrupt();
			throw new PoolObjectReleasingInterruptedException(key, objectToRelease, e);
		}
		
		if (key == null)
			throw new InvalidPoolOperationException(null, objectToRelease, _MESSAGE_RELEASING_OF_STRANGER);
			
		_pool.release(key, objectToRelease);
	}
	
	@Override
	public void close() throws Exception {
		_pool.close();
	}
	
	public TK getKeyByObject(TV obtainedObject) throws InterruptedException {

		_obtainedObjectToKeyLock.lockRead();
		
		try {			
			return _obtainedObjectToItsKey.get(obtainedObject);
		} finally {
			_obtainedObjectToKeyLock.unlockRead();
		}
	}
	
	private void waitSafely(int secondsNumberToWait, TK requestedKey) throws PoolObjectObtainingInterruptedException {
		
		try {
			Thread.sleep(secondsNumberToWait * 1000);
		} catch (InterruptedException e) {
			Thread.currentThread().interrupt();
			throw new PoolObjectObtainingInterruptedException(requestedKey, e);
		}
	}
	
	private void rememberObtainedObject(TK key, TV obtainedObject) throws PoolObjectObtainingInterruptedException {
		
		try {
			if (getKeyByObject(obtainedObject) != null)
				return;
			
			_obtainedObjectToKeyLock.lockWrite();			
			try {				
				_obtainedObjectToItsKey.put(obtainedObject, key);
			} finally {
				_obtainedObjectToKeyLock.unlockWrite();
			}
		} catch (InterruptedException e) {
			Thread.currentThread().interrupt();
			
			Throwable cause = e;
			String message = String.format(_MESSAGE_PATTERN_OBTAINING_INTERRUPTED, key);
			
			try {
				_pool.release(key, obtainedObject);
			} catch (InvalidPoolOperationException | PoolObjectReleasingInterruptedException ee) {
				cause = ee;
				message += ". Releasing also hasn't finished. Look at cause for details";
			}
			
			throw new PoolObjectObtainingInterruptedException(key, message, cause);
		}
	}
}
